src/providers/files/files_init.c

```diff
@@ -23,39 +23,190 @@
+                if (passwd_files[i] == NULL) {
+                    ret = ENOMEM;
+                    goto done;
+                }
+        }
+    }
+
+    /* Default to /etc/group standard */
+    if (conf_group_files == NULL) {
+        num_group_files = 1;
+
+        group_files = talloc_zero_array(tmp_ctx, const char *, 2);
+        if (group_files == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "talloc_zero_array() failed\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC,
+              "Using default group file: [%s].\n", DEFAULT_GROUP_FILE);
+        group_files[0] = talloc_strdup(group_files, DEFAULT_GROUP_FILE);
+        if (group_files[0] == NULL) {
+            ret = ENOMEM;
+            goto done;
+        }
+    /* Retrieve list of group files */
+    } else {
+        ret = split_on_separator(tmp_ctx, conf_group_files, ',', true, true,
+                                 &group_list, &num_group_files);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Failed to parse group files!\n");
+            goto done;
+        }
+
+        group_files = talloc_zero_array(tmp_ctx, const char *,
+                                         num_group_files + 1);
+        if (group_files == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "talloc_zero_array() failed\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        for (i = 0; i < num_group_files; i++) {
+            DEBUG(SSSDBG_TRACE_FUNC,
+                  "Using group file: [%s].\n", group_list[i]);
+            group_files[i] = talloc_strdup(group_files, group_list[i]);
+                if (group_files[i] == NULL) {
+                    ret = ENOMEM;
+                    goto done;
+                }
+        }
+    }
+
+    *_passwd_files = talloc_steal(mem_ctx, passwd_files);
+    *_group_files = talloc_steal(mem_ctx, group_files);
+    *_num_passwd_files = num_passwd_files;
+    *_num_group_files = num_group_files;
+
+    ret = EOK;
+
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
 int sssm_files_init(TALLOC_CTX *mem_ctx,
                     struct be_ctx *be_ctx,
                     struct data_provider *provider,
                     const char *module_name,
                     void **_module_data)
 {
     struct files_id_ctx *ctx;
-    int ret;
-    const char *passwd_file = NULL;
-    const char *group_file = NULL;
-
-    /* So far this is mostly useful for tests */
-    passwd_file = getenv("SSS_FILES_PASSWD");
-    if (passwd_file == NULL) {
-        passwd_file = "/etc/passwd";
-    }
+    const char **passwd_files = NULL;
+    const char **group_files = NULL;
+    int num_passwd_files = 0;
+    int num_group_files = 0;
+    errno_t ret;
 
-    group_file = getenv("SSS_FILES_GROUP");
-    if (group_file == NULL) {
-        group_file = "/etc/group";
-    }
+    ret = files_init_file_sources(mem_ctx, be_ctx, &passwd_files,
```

Comment:
We don't check ret for failures here, we probably should.

We should probably move this call after allocating the ctx and use ctx in place of mem_ctx here.

Reply:
Yes, good catch.